cupy/cuda/cupy_thrust.cu

```diff
@@ -43,6 +44,105 @@ public:
 };
 
 
+/* ------------------------------------ Minimum boilerplate to support complex numbers ------------------------------------ */
+// We need a specialized operator< here in order to match the NumPy behavior:
+// "The sort order for complex numbers is lexicographic. If both the real and imaginary parts are non-nan then the order is
+// determined by the real parts except when they are equal, in which case the order is determined by the imaginary parts.
+//
+// In numpy versions >= 1.4.0 nan values are sorted to the end. The extended sort order is:
+//     Real: [R, nan]
+//     Complex: [R + Rj, R + nanj, nan + Rj, nan + nanj]
+// where R is a non-nan real value. Complex values with the same nan placements are sorted according to the non-nan part if
+// it exists. Non-nan values are sorted as before."
+// Ref: https://docs.scipy.org/doc/numpy/reference/generated/numpy.sort.html
+
+template <typename T>
+__host__ __device__ __forceinline__ int _check_nan(const T& lhs, const T& rhs) {
+    // return values: +1: lhs < rhs; 0: neither is NaN; -1: lhs > rhs
+
+    bool lhsRe = isnan(lhs.x);
+    bool lhsIm = isnan(lhs.y);
+    bool rhsRe = isnan(rhs.x);
+    bool rhsIm = isnan(rhs.y);
+
+    if (!lhsRe && !lhsIm && !rhsRe && !rhsIm)
+        return 0;
+    if (!lhsRe && !lhsIm && (rhsRe || rhsIm))
+        return 1;
+    if ((lhsRe || lhsIm) && !rhsRe && !rhsIm)
+        return -1;
+
+    // at this point, we know both lhs and rhs have at least one NaN
+    if (lhsRe && !rhsRe)
+        return -1;
+    if (!lhsRe && rhsRe)
+        return 1;
+    if (lhsIm && !rhsIm)
+        return -1;
+    if (!lhsIm && rhsIm)
+        return 1;
```

Comment:
Would it be possible to have two ifs?
```c
if (lhsRe && !rhsRe || lhsIm && !rhsIm) {
    return -1;
}
if (!lhsRe && rhsRe || !lhsIm && rhsIm) {
    return 1;
}
```

Also please use curly braces

Reply:
Unfortunately, this doesn't work for this case:
```
0.5 + nan I < nan - 0.7 I
```
as the first `if` would make a wrong comparison (make a judgement too early). If you want an explicit comparison, we will need to compare both real and imaginary parts at the same time, something like
```
// for comparing a + nan I < nan + b I
if ((!lhsRe && rhsRe) && (lhsIm && !rhsIm)) {
    return true;  // or 1 if returning int
}